File: letter-case-permutation.cpp
class Solution {
public:
    vector<string> answer;
    void backt(string S, int begin) {
        if (begin == S.size()) {
            answer.push_back(S);
        } else {
            if (isalpha(S[begin])) {
                S[begin] = toupper(S[begin]);
                backt(S, begin+1);
                S[begin] = tolower(S[begin]);
                backt(S, begin+1);
                
            } else {
                backt(S, begin+1);
            }
            
        }
        
    }
    vector<string> letterCasePermutation(string S) {
        backt(S, 0);
        return answer;
        
    }
};

File: letterCasePermutation.py
class Solution(object):
    def letterCasePermutation(self, s):
        output = set()
        def backtrack(st, index):
            if index >= len(s):
                output.add(st)
                return
            if st[index].isalpha():
                if ord(st[index]) <= 90:
                    backtrack(st[:index] + chr(ord(st[index]) + 32) + st[index+1:], index+1)
                else:
                    backtrack(st[:index] + chr(ord(st[index]) - 32) + st[index+1:], index+1)
            backtrack(st, index+1)
        backtrack(s, 0)
        return output
        
                
